### PR TITLE
Feature/tokenomics

### PR DESCRIPTION
**Pull request by Jayashsatolia403**

### CODE REVIEW

File: .github/workflows/tokenomics_ci.yml [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+          echo "TOKENOMICS_MIN_STAKE_TESTS=$(echo $(([ -z 'TestMinStakeForProviders' ] && echo '') || echo 'TestMinStakeForProviders'))" >> $GITHUB_ENV
+          echo "CURRENT_BRANCH=$(echo $(([ -z '${{ github.head_ref }}' ] && echo  ${GITHUB_REF#refs/*/}) || echo '${{ github.head_ref }}'))" >> $GITHUB_ENV
+          echo "REPO_SNAPSHOTS_BRANCH=$(echo $(([ -z '${{github.event.inputs.repo_snapshots_branch}}' ] && echo '<CURRENT BRANCH WITH FALLBACK TO CURRENT_SPRINT>') || echo '${{github.event.inputs.repo_snapshots_branch}}'))" >> $GITHUB_ENV
+
+
+      - name: "Deploy 0Chain"
+        if: github.event_name == 'push' || github.event.inputs.existing_network == ''
+        uses: 0chain/actions/deploy-0chain-minstake@feature/tokenomics
+        with:
+          repo_snapshots_branch: fix/minstake
```
Comment (Kishan-Dhakan) [inline .github/workflows/tokenomics_ci.yml:95]: revert to current-sprint

File: .github/workflows/tokenomics_ci.yml [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+          TENDERLY_FORK_ID: ${{ secrets.TENDERLY_FORK_ID }}
+          graphnode_sc: ${{ secrets.GRAPHNODE_SC }}
+          graphnode_network: ${{ secrets.GRAPHNODE_NETWORK }}
+          graphnode_ethereum_node_url: https://rpc.tenderly.co/fork/${{ secrets.TENDERLY_FORK_ID }}
+          svc_account_secret: ${{ secrets.SVC_ACCOUNT_SECRET }}
+
+      - name: "Run Min Stake System tests"
+        uses: 0chain/actions/run-system-tests-tokenomics@feature/tokenomics
+        with:
+          repo_snapshots_branch: fix/minstake
```
Comment (Kishan-Dhakan) [inline .github/workflows/tokenomics_ci.yml:108]: revert to current-sprint (all instances please)

File: .github/workflows/tokenomics_ci.yml [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+          tokenomics_test_filter: ${{ env.TOKENOMICS_MIN_STAKE_TESTS }}
+          TENDERLY_FORK_ID: ${{ secrets.TENDERLY_FORK_ID }}
+          run_smoke_tests: ${{ inputs.run_smoke_tests }}
+          S3_ACCESS_KEY: ${{ secrets.S3_ACCESS_KEY }}
+          S3_SECRET_KEY: ${{ secrets.S3_SECRET_KEY }}
+
+
+      - name: "Deploy 0Chain"
+        if: github.event_name == 'push' || github.event.inputs.existing_network == ''
+        uses: 0chain/actions/deploy-0chain-2b2v@feature/tokenomics
```
Comment (Kishan-Dhakan) [inline .github/workflows/tokenomics_ci.yml:127]: revert @feature/tokenomics to @master (all instances, actions is merged)

File: tests/tokenomics_tests/config/wallets/UTC--2022-11-24T20-45-27.508229000Z--c49926c4124cee1cba0ea94ea31a6c12318df947 [hunk 4]
```diff
@@ -0,0 +1 @@
+{"address":"c49926c4124cee1cba0ea94ea31a6c12318df947","crypto":{"cipher":"aes-128-ctr","ciphertext":"d4836bccc34c9ae4a968ec39b380644a9889e8327888f1604e66c320fec48bef","cipherparams":{"iv":"c4fdfd5238be9481e773aec3d97b030c"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":262144,"p":1,"r":8,"salt":"278e0fbc2f8f8a07273954494c77b196331b6bbe88f2d7ba7f1c15b6b7b52d6d"},"mac":"4e50efbbe89f009dc2b400e00ad07e6e13a1e882848094d95f6a43326904456e"},"id":"ae56a19b-f04d-46b6-b3a8-25827bd67f25","version":3}
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/config/wallets/UTC--2022-11-24T20-45-27.508229000Z--c49926c4124cee1cba0ea94ea31a6c12318df947:1]: Can we remove these temp files?

**Review by Kishan-Dhakan — CHANGES_REQUESTED** [review]

File: .github/workflows/tokenomics_ci.yml [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+          repository: "0chain/system_test"
+          sha: master
+      - name: Success
+        if: ${{ env.BUILD_STATE == 'success' }}
+        run: echo "master is successful!"
+      - name: Not Successful
+        if: ${{ env.BUILD_STATE != 'success' }}
+        run: |
+          echo "master is NOT successful! was: [${{ env.BUILD_STATE }}]"
+          exit 1
```
Comment (Kishan-Dhakan) [inline .github/workflows/tokenomics_ci.yml]: can remove this check, it's not relevant for this cron

File: tests/tokenomics_tests/allocation_test.go [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+	var validatorListString []string
+	for _, validator := range validatorList {
+		validatorListString = append(validatorListString, validator.ID)
+	}
+
+	stakeTokensToBlobbersAndValidators(t, blobberListString, validatorListString, configPath, []float64{
+		1, 1, 1, 1,
+	}, 1)
+
+	t.RunSequentiallyWithTimeout("Create + Upload + Upgrade equal read price 0.1", (500*time.Minute)+(40*time.Second), func(t *test.SystemTest) {
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/allocation_test.go:77]: Can you turn these 500 minute timeouts (all instances) into shorter periods, even 20 or 30 minutes is fine we just want to know if test got stuck or performance is severely affected anywhere

File: tests/tokenomics_tests/allocation_test.go [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+}
+
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/allocation_test.go:692]: You should use strings.Contains

File: tests/tokenomics_tests/min_stake_test.go [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+	ChallengePassReward
+	ChallengeSlashPenalty
+	CancellationChargeReward
+	NumOfRewards
+)
+
+func TestMinStakeForProviders(testSetup *testing.T) {
+	t := test.NewSystemTest(testSetup)
+
+	t.TestSetup("set storage config to use time_unit as 10 minutes", func() {
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/min_stake_test.go:41]: `config to use time_unit as 30 minutes`

File: tests/tokenomics_tests/blobber_challenge_reward_test.go [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+	AllocationID   string           `json:"allocation_id"`
+	BlobberID      string           `json:"blobber_id"`
+	ValidatorsID   string           `json:"validators_id"`
+	Seed           int64            `json:"seed"`
+	AllocationRoot string           `json:"allocation_root"`
+	Responded      int64            `json:"responded"`
+	Passed         bool             `json:"passed"`
+	RoundResponded int64            `json:"round_responded"`
+	ExpiredN       int              `json:"expired_n"`
+}
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/blobber_challenge_reward_test.go:304]: Can move to reusable structs to internal/cli/model and import from there as a best practice

File: tests/tokenomics_tests/blobber_challenge_reward_test.go [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+
+		fileNames = append(fileNames, filename)
+	}
+
+	for i := 0; i < numDeletes; i++ {
+		remotepath := "/dir/"
+		filename := fileNames[numFiles-i-1]
+
+		output, err := utils.DeleteFile(t, utils.EscapedTestName(t), utils.CreateParams(map[string]interface{}{
+			// fetch the latest block in the chain
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/blobber_challenge_reward_test.go:398]: What is this comment saying?

File: tests/tokenomics_tests/blobber_read_test.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestBlobberReadReward(testSetup *testing.T) {
+	t := test.NewSystemTest(testSetup)
+
+	output, err := utils.UpdateStorageSCConfig(t, scOwnerWallet, map[string]string{
+		"time_unit": "10m",
+	}, true)
+	require.Nil(t, err, strings.Join(output, "\n"))
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/blobber_read_test.go:27]: Can you move it to TestSetup and add cleanup like in other tests please

File: tests/tokenomics_tests/blobber_slash_penalty_test.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestBlobberSlashPenalty(testSetup *testing.T) {
+	t := test.NewSystemTest(testSetup)
+
+	output, err := utils.UpdateStorageSCConfig(t, scOwnerWallet, map[string]string{
+		"time_unit": "20m",
+	}, true)
+	require.Nil(t, err, strings.Join(output, "\n"))
```
Comment (Kishan-Dhakan) [inline tests/tokenomics_tests/blobber_slash_penalty_test.go:24]: Move to testsetup (and cleanup if needed) please

**Review by Kishan-Dhakan — CHANGES_REQUESTED** [review]
Minor comments in the tests themselves. Some are just best practices and can be done later, others like moving assertions to TestSetup would need to be done

File: tests/tokenomics_tests/allocation_test.go (hunk 7 above)
Comment (Jayashsatolia403) [reply tests/tokenomics_tests/allocation_test.go:692]: We can not use strings.Contains to search string in list of strings. It is to check substring in a string I think.

Comment (Kishan-Dhakan) [reply tests/tokenomics_tests/allocation_test.go:692]: Woops, my bad. Misread the code

**Pull request merged**